api4/upload.go

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package api4
+
+import (
+	"net/http"
+
+	"github.com/mattermost/mattermost-server/v5/audit"
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+func (api *API) InitUpload() {
+	api.BaseRoutes.Uploads.Handle("", api.ApiSessionRequired(createUpload)).Methods("POST")
+	api.BaseRoutes.Upload.Handle("", api.ApiSessionRequired(getUpload)).Methods("GET")
+	api.BaseRoutes.Upload.Handle("", api.ApiSessionRequired(uploadData)).Methods("POST")
```

Comment:
why new endpoints?

Reply:
For a couple of reasons mainly:

1. To support this feature I've introduced a new `UploadSession` entity  which is separate from `FileInfo` and it made sense for it to have its own endpoint.
2. The `/files` endpoint (POST method) is already polluted with two different ways of uploading files (multipart and file in body). Moreover, both suffer from a problem we are trying to eliminate which is the loading of whole files in memory. Since the required changes are non trivial and possibly breaking, I've thought that implementing a totally new codepath would be the fastest, cleanest and most reliable way to both solve this issue and at the same time implement the requested feature. The idea is that once this has been thoroughly tested and successfully implemented on the front-end we can gradually move away from the previous calls through deprecation.